# Housing Rental Analysis for San Francisco

In this challenge, your job is to use your data visualisation skills, including aggregation, interactive visualisations, and geospatial analysis, to find properties in the San Francisco market that are viable investment opportunities.

### Instructions

Use the `san_francisco_housing.ipynb` notebook to visualise and analyse the real-estate data.

Note that this assignment requires you to create a visualisation by using hvPlot and GeoViews. Additionally, you need to read the `sfo_neighborhoods_census_data.csv` file from the `Resources` folder into the notebook and create the DataFrame that you’ll use in the analysis.

The main task in this Challenge is to visualise and analyse the real-estate data in your Jupyter notebook. Use the `san_francisco_housing.ipynb` notebook to complete the following tasks:

* Calculate and plot the housing units per year.

* Calculate and plot the average prices per square foot.

* Compare the average prices by neighbourhood.

* Build an interactive neighbourhood map.

* Compose your data story.

#### Calculate and Plot the Housing Units per Year

For this part of the assignment, use numerical and visual aggregation to calculate the number of housing units per year, and then visualise the results as a bar chart. To do so, complete the following steps:

1. Use the `groupby` function to group the data by year. Aggregate the results by the `mean` of the groups.

2. Use the `hvplot` function to plot the `housing_units_by_year` DataFrame as a bar chart. Make the x-axis represent the `year` and the y-axis represent the `housing_units`.

3. Style and format the line plot to ensure a professionally styled visualisation.

4. Note that your resulting plot should appear similar to the following image:

    ![A screenshot depicts an example of the resulting bar chart.](Images/6-4-zoomed-housing-units-by-year.png)

5. Answer the following question:

    * What’s the overall trend in housing units over the period that you’re analysing?

#### Calculate and Plot the Average Sale Prices per Square Foot

For this part of the assignment, use numerical and visual aggregation to calculate the average prices per square foot, and then visualise the results as a bar chart. To do so, complete the following steps:

1. Group the data by year, and then average the results. What’s the lowest gross rent that’s reported for the years that the DataFrame includes?

2. Create a new DataFrame named `prices_square_foot_by_year` by filtering out the “housing_units” column. The new DataFrame should include the averages per year for only the sale price per square foot and the gross rent.

3. Use hvPlot to plot the `prices_square_foot_by_year` DataFrame as a line plot.

    > **Hint** This single plot will include lines for both `sale_price_sqr_foot` and `gross_rent`.

4. Style and format the line plot to ensure a professionally styled visualisation.

5. Note that your resulting plot should appear similar to the following image:

    ![A screenshot depicts an example of the resulting plot.](Images/6-4-avg-sale-px-sq-foot-gross-rent.png)

6. Use both the `prices_square_foot_by_year` DataFrame and interactive plots to answer the following questions:

    * Did any year experience a drop in the average sale price per square foot compared to the previous year?

    * If so, did the gross rent increase or decrease during that year?

#### Compare the Average Sale Prices by Neighbourhood

For this part of the assignment, use interactive visualisations and widgets to explore the average sale price per square foot by neighbourhood. To do so, complete the following steps:

1. Create a new DataFrame that groups the original DataFrame by year and neighbourhood. Aggregate the results by the `mean` of the groups.

2. Filter out the “housing_units” column to create a DataFrame that includes only the `sale_price_sqr_foot` and `gross_rent` averages per year.

3. Create an interactive line plot with hvPlot that visualises both `sale_price_sqr_foot` and `gross_rent`. Set the x-axis parameter to the year (`x="year"`). Use the `groupby` parameter to create an interactive widget for `neighbourhood`.

4. Style and format the line plot to ensure a professionally styled visualisation.

5. Note that your resulting plot should appear similar to the following image:

    ![A screenshot depicts an example of the resulting plot.](Images/6-4-pricing-info-by-neighborhood.png)

6. Use the interactive visualisation to answer the following question:

    * For the Anza Vista neighbourhood, is the average sale price per square foot for 2016 more or less than the price that’s listed for 2012?

#### Build an Interactive Neighbourhood Map

For this part of the assignment, explore the geospatial relationships in the data by using interactive visualisations with hvPlot and GeoViews. To build your map, use the `sfo_data_df` DataFrame (created during the initial import), which includes the neighbourhood location data with the average prices. To do all this, complete the following steps:

1. Read the `neighbourhood_coordinates.csv` file from the `Resources` folder into the notebook, and create a DataFrame named `neighbourhood_locations_df`. Be sure to set the `index_col` of the DataFrame as “Neighbourhood”.

2. Using the original `sfo_data_df` Dataframe, create a DataFrame named `all_neighbourhood_info_df` that groups the data by neighbourhood. Aggregate the results by the `mean` of the group.

3. Review the two code cells that concatenate the `neighbourhood_locations_df` DataFrame with the `all_neighbourhood_info_df` DataFrame. Note that the first cell uses the [Pandas concat function](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.concat.html) to create a DataFrame named `all_neighbourhoods_df`. The second cell cleans the data and sets the “Neighbourhood” column. Be sure to run these cells to create the `all_neighbourhoods_df` DataFrame, which you’ll need to create the geospatial visualisation.

4. Using hvPlot with GeoViews enabled, create a `points` plot for the `all_neighbourhoods_df` DataFrame. Be sure to do the following:

    * Set the `size` parameter to “sale_price_sqr_foot”.

    * Set the `color` parameter to “gross_rent”.

    * Set the `frame_width` parameter to 700.

    * Set the `frame_height` parameter to 500.

    * Include a descriptive title.

    Note that your resulting plot should appear similar to the following image:

    ![A screenshot depicts an example of a scatter plot created with hvPlot and GeoViews.](Images/6-4-geoviews-plot.png)

5. Use the interactive map to answer the following question:

    * Which neighbourhood has the highest gross rent, and which has the highest sale price per square foot?

#### Compose Your Data Story

Based on the visualisations that you created, answer the following questions:

* How does the trend in rental income growth compare to the trend in sales prices? Does this same trend hold true for all the neighbourhoods across San Francisco?

* What insights can you share with your company about the potential one-click, buy-and-rent strategy that they're pursuing? Do neighbourhoods exist that you would suggest for investment, and why?

In [2]:
# Import the required libraries and dependencies
import pandas as pd
import hvplot.pandas
from pathlib import Path

## Import the data 

In [13]:
# Using the read_csv function and Path module, create a DataFrame 
# by importing the sfo_neighbourhoods_census_data.csv file from the Resources folder
sfo_data_df = pd.read_csv('./Resources/sfo_neighbourhoods_census_data.csv')
sfo_data_df.head(20)

# Review the first and last five rows of the DataFrame
# YOUR CODE HERE
# YOUR CODE HERE

year                  neighbourhood  sale_price_sqr_foot  housing_units  \
0   2010                   Alamo Square           291.182945         372560   
1   2010                     Anza Vista           267.932583         372560   
2   2010                        Bayview           170.098665         372560   
3   2010               Buena Vista Park           347.394919         372560   
4   2010               Central Richmond           319.027623         372560   
5   2010                 Central Sunset           418.172493         372560   
6   2010                 Corona Heights           369.359338         372560   
7   2010                     Cow Hollow           569.379968         372560   
8   2010                  Croker Amazon           165.645730         372560   
9   2010                Diamond Heights           456.930822         372560   
10  2010                      Downtown            241.804552         372560   
11  2010  Eureka Valley/Dolores Heights           403.118313         372560   
12  2010                      Excelsior           364.928336         372560   
13  2010       Financial District North           401.749967         372560   
14  2010       Financial District South           162.821210         372560   
15  2010                  Forest Knolls           209.049327         372560   
16  2010                      Glen Park           715.586438         372560   
17  2010            Golden Gate Heights           551.801119         372560   
18  2010                 Haight Ashbury           254.473779         372560   
19  2010                   Hayes Valley           272.284980         372560   

    gross_rent  
0         1239  
1         1239  
2         1239  
3         1239  
4         1239  
5         1239  
6         1239  
7         1239  
8         1239  
9         1239  
10        1239  
11        1239  
12        1239  
13        1239  
14        1239  
15        1239  
16        1239  
17        1239  
18        1239  
19        1239

---

## Calculate and Plot the Housing Units per Year

For this part of the assignment, use numerical and visual aggregation to calculate the number of housing units per year, and then visualise the results as a bar chart. To do so, complete the following steps:

1. Use the `groupby` function to group the data by year. Aggregate the results by the `mean` of the groups.

2. Use the `hvplot` function to plot the `housing_units_by_year` DataFrame as a bar chart. Make the x-axis represent the `year` and the y-axis represent the `housing_units`.

3. Style and format the line plot to ensure a professionally styled visualisation.

4. Note that your resulting plot should appear similar to the following image:

    ![A screenshot depicts an example of the resulting bar chart.](Images/6-4-zoomed-housing-units-by-year.png)

5. Answer the following question:

    * What’s the overall trend in housing units over the period that you’re analysing?

### Step 1: Use the `groupby` function to group the data by year. Aggregate the results by the `mean` of the groups.

In [20]:
# Create a numerical aggregation that groups the data by the year and then averages the results.
housing_units_by_year = sfo_data_df.drop('neighbourhood', axis = 1).groupby('year').mean()

housing_units_by_year.head(20)

# Review the DataFrame
# YOUR CODE HERE

sale_price_sqr_foot  housing_units  gross_rent
year                                                
2010           369.344353       372560.0      1239.0
2011           341.903429       374507.0      1530.0
2012           399.389968       376454.0      2324.0
2013           483.600304       378401.0      2971.0
2014           556.277273       380348.0      3528.0
2015           632.540352       382295.0      3739.0
2016           697.643709       384242.0      4390.0

### Step 2: Use the `hvplot` function to plot the `housing_units_by_year` DataFrame as a bar chart. Make the x-axis represent the `year` and the y-axis represent the `housing_units`.

### Step 3: Style and format the line plot to ensure a professionally styled visualisation.

In [36]:
# Create a visual aggregation explore the housing units by year
housing_units_by_year.hvplot.bar(
    x = 'year',
    y = 'housing_units').opts(
                        yformatter = '%.0f', 
                        ylim = (360000, 390000), 
                        xlabel = 'Year', 
                        ylabel = 'Housing Units',
                        color = 'blue',
                        title = 'Housing units in San Francisco between 2010 and 2016')
                        

:Bars   [year]   (housing_units)

### Step 5: Answer the following question:

**Question:** What is the overall trend in housing_units over the period being analysed?

**Answer:** The quantity of units on the market experiences steady growth over the 7 year period.

---

## Calculate and Plot the Average Sale Prices per Square Foot

For this part of the assignment, use numerical and visual aggregation to calculate the average prices per square foot, and then visualise the results as a bar chart. To do so, complete the following steps:

1. Group the data by year, and then average the results. What’s the lowest gross rent that’s reported for the years that the DataFrame includes?

2. Create a new DataFrame named `prices_square_foot_by_year` by filtering out the “housing_units” column. The new DataFrame should include the averages per year for only the sale price per square foot and the gross rent.

3. Use hvPlot to plot the `prices_square_foot_by_year` DataFrame as a line plot.

    > **Hint** This single plot will include lines for both `sale_price_sqr_foot` and `gross_rent`.

4. Style and format the line plot to ensure a professionally styled visualisation.

5. Note that your resulting plot should appear similar to the following image:

    ![A screenshot depicts an example of the resulting plot.](Images/6-4-avg-sale-px-sq-foot-gross-rent.png)

6. Use both the `prices_square_foot_by_year` DataFrame and interactive plots to answer the following questions:

    * Did any year experience a drop in the average sale price per square foot compared to the previous year?

    * If so, did the gross rent increase or decrease during that year?

### Step 1: Group the data by year, and then average the results.

In [47]:
# Create a numerical aggregation by grouping the data by year and averaging the results
prices_square_foot_by_year = sfo_data_df[['year', 'sale_price_sqr_foot', 'gross_rent']]
prices_square_foot_by_year = prices_square_foot_by_year.groupby('year').mean()

# Review the resulting DataFrame
prices_square_foot_by_year.head(20)

sale_price_sqr_foot  gross_rent
year                                 
2010           369.344353      1239.0
2011           341.903429      1530.0
2012           399.389968      2324.0
2013           483.600304      2971.0
2014           556.277273      3528.0
2015           632.540352      3739.0
2016           697.643709      4390.0

**Question:** What is the lowest gross rent reported for the years included in the DataFrame?


**Answer:** The lowest Gross rent happened during the year 2010

### Step 2: Create a new DataFrame named `prices_square_foot_by_year` by filtering out the “housing_units” column. The new DataFrame should include the averages per year for only the sale price per square foot and the gross rent.

In [ ]:
# Not sure why we need this step, since all the data has been already displayed in the cell above.

# Filter out the housing_units column, creating a new DataFrame 
# Keep only sale_price_sqr_foot and gross_rent averages per year
#prices_square_foot_by_year = # YOUR CODE HERE

# Review the DataFrame
# YOUR CODE HERE



### Step 3: Use hvPlot to plot the `prices_square_foot_by_year` DataFrame as a line plot.

> **Hint** This single plot will include lines for both `sale_price_sqr_foot` and `gross_rent`

### Step 4: Style and format the line plot to ensure a professionally styled visualisation.


In [108]:
# Plot prices_square_foot_by_year. 
# Inclued labels for the x- and y-axes, and a title.

prices_square_foot_by_year.hvplot(title = 'Gross Rent and Sale price per square foot for San Francisco 2010-2016',
                                 xlabel = 'Year',
                                 ylabel = 'Sales price per square foor/Gross Rent',
                                 width = 900,
                                 height = 500 ,
                                 grid = True
                                 )


#OPTIONAL CODE with color modifications.
# |
# |
# |
# \/

# prices_square_foot_by_year_plot = prices_square_foot_by_year.hvplot(x = 'year',
#                                                                     y = 'sale_price_sqr_foot',
#                                                                     legend = True,
#                                                                     label='Prices per Square Foot'
#                                                                    ).opts(color = 'green',
#                                                                           xlabel = 'Year',
#                                                                           ylabel = 'Price per square foot'
#                                                                          )
# gross_rent_plot = prices_square_foot_by_year.hvplot(x ='year',
#                                                     y ='gross_rent',
#                                                    legend = 'right',
#                                                    label='Gross Rent').opts(color = 'black',
#                                                                           xlabel = 'Year',
#                                                                           ylabel = 'Gross Rent')



# combined_plot = (prices_square_foot_by_year_plot * gross_rent_plot).opts(title = 'Gross Rent and Sale price per square foot',
#                                                                         width = 900,
#                                                                         height = 500,
#                                                                         ylabel = 'Gross Rent/Price per square foot',
#                                                                         show_grid = True
#                                                                         )
# gross_rent_plot
# combined_plot

:NdOverlay   [Variable]
   :Curve   [year]   (value)

### Step 6: Use both the `prices_square_foot_by_year` DataFrame and interactive plots to answer the following questions:

**Question:** Did any year experience a drop in the average sale price per square foot compared to the previous year?

**Answer:** There was a minor sales price drop in 2011.


**Question:** If so, did the gross rent increase or decrease during that year?

**Answer:** Despite the reduce sales price, the rental price still increased that year

---

## Compare the Average Sale Prices by Neighbourhood

For this part of the assignment, use interactive visualisations and widgets to explore the average sale price per square foot by neighbourhood. To do so, complete the following steps:

1. Create a new DataFrame that groups the original DataFrame by year and neighbourhood. Aggregate the results by the `mean` of the groups.

2. Filter out the “housing_units” column to create a DataFrame that includes only the `sale_price_sqr_foot` and `gross_rent` averages per year.

3. Create an interactive line plot with hvPlot that visualises both `sale_price_sqr_foot` and `gross_rent`. Set the x-axis parameter to the year (`x="year"`). Use the `groupby` parameter to create an interactive widget for `neighbourhood`.

4. Style and format the line plot to ensure a professionally styled visualisation.

5. Note that your resulting plot should appear similar to the following image:

    ![A screenshot depicts an example of the resulting plot.](Images/6-4-pricing-info-by-neighborhood.png)

6. Use the interactive visualisation to answer the following question:

    * For the Anza Vista neighbourhood, is the average sale price per square foot for 2016 more or less than the price that’s listed for 2012?


### Step 1: Create a new DataFrame that groups the original DataFrame by year and neighbourhood. Aggregate the results by the `mean` of the groups.

In [112]:
# Group by year and neighbourhood and then create a new dataframe of the mean values
prices_by_year_by_neighbourhood = sfo_data_df.groupby(['year', 'neighbourhood']).mean()


# Review the DataFrame
prices_by_year_by_neighbourhood.tail(50)

sale_price_sqr_foot  housing_units  \
year neighbourhood                                                       
2015 Van Ness/ Civic Center                  677.838232       382295.0   
     Visitacion Valley                       302.221873       382295.0   
     Western Addition                        481.469244       382295.0   
     Yerba Buena                             963.522606       382295.0   
2016 Alamo Square                            340.981835       384242.0   
     Anza Vista                               88.401511       384242.0   
     Bernal Heights                          986.842105       384242.0   
     Buena Vista Park                        851.401405       384242.0   
     Central Richmond                        661.064426       384242.0   
     Central Sunset                          581.370875       384242.0   
     Cow Hollow                             1059.065602       384242.0   
     Croker Amazon                           566.701176       384242.0   
     Downtown                                718.776830       384242.0   
     Eureka Valley/Dolores Heights          1026.791108       384242.0   
     Excelsior                               534.226250       384242.0   
     Financial District North                       NaN       384242.0   
     Glen Park                               945.566502       384242.0   
     Haight Ashbury                          531.835032       384242.0   
     Hayes Valley                            175.066069       384242.0   
     Ingleside Heights                       571.579983       384242.0   
     Inner Mission                           647.770447       384242.0   
     Inner Parkside                          518.112650       384242.0   
     Inner Richmond                          403.311132       384242.0   
     Inner Sunset                            445.923828       384242.0   
     Jordan Park/Laurel Heights              704.566419       384242.0   
     Lower Pacific Heights                   636.430770       384242.0   
     Marina                                  431.944602       384242.0   
     Merced Heights                         1416.666667       384242.0   
     Mission Bay                             924.569116       384242.0   
     Mission Terrace                         879.017013       384242.0   
     Nob Hill                                777.008315       384242.0   
     Noe Valley                              799.495651       384242.0   
     North Beach                             407.553322       384242.0   
     Outer Mission                            40.772625       384242.0   
     Outer Parkside                          732.720684       384242.0   
     Outer Richmond                         1196.645437       384242.0   
     Outer Sunset                            510.576907       384242.0   
     Pacific Heights                         996.389364       384242.0   
     Park North                              211.301370       384242.0   
     Parnassus/Ashbury Heights              1207.997485       384242.0   
     Potrero Hill                            707.651609       384242.0   
     Presidio Heights                       1465.968586       384242.0   
     Russian Hill                           1024.339877       384242.0   
     South of Market                         531.136374       384242.0   
     Sunnyside                               751.173709       384242.0   
     Telegraph Hill                          903.049771       384242.0   
     Twin Peaks                              970.085470       384242.0   
     Van Ness/ Civic Center                  552.602567       384242.0   
     Visitacion Valley                       328.319007       384242.0   
     Westwood Park                           631.195426       384242.0   

                                    gross_rent  
year neighbourhood                              
2015 Van Ness/ Civic Center             3739.0  
     Visitacion Valley                  

### Step 2: Filter out the “housing_units” column to create a DataFrame that includes only the `sale_price_sqr_foot` and `gross_rent` averages per year.

In [115]:
# Filter out the housing_units
prices_by_year_by_neighbourhood = prices_by_year_by_neighbourhood[['sale_price_sqr_foot', 'gross_rent']]



# Review the first and last five rows of the DataFrame
print (prices_by_year_by_neighbourhood.head(100))
print (prices_by_year_by_neighbourhood.tail(100))

                       sale_price_sqr_foot  gross_rent
year neighbourhood                                    
2010 Alamo Square               291.182945      1239.0
     Anza Vista                 267.932583      1239.0
     Bayview                    170.098665      1239.0
     Buena Vista Park           347.394919      1239.0
     Central Richmond           319.027623      1239.0
...                                    ...         ...
2011 North Waterfront           346.961160      1530.0
     Outer Mission              142.142568      1530.0
     Outer Parkside             241.945656      1530.0
     Outer Richmond             271.090140      1530.0
     Outer Sunset               297.245558      1530.0

[100 rows x 2 columns]
                             sale_price_sqr_foot  gross_rent
year neighbourhood                                          
2015 Buena Vista Park                 551.695622      3739.0
     Central Richmond                 473.163815      3739.0
     Central Suns

### Step 3: Create an interactive line plot with hvPlot that visualises both `sale_price_sqr_foot` and `gross_rent`. Set the x-axis parameter to the year (`x="year"`). Use the `groupby` parameter to create an interactive widget for `neighbourhood`.

### Step 4: Style and format the line plot to ensure a professionally styled visualisation.

In [120]:
# Use hvplot to create an interactive line plot of the average price per square foot
# The plot should have a dropdown selector for the neighbourhood
prices_by_year_by_neighbourhood.hvplot(x = 'year', 
                                       groupby = 'neighbourhood', 
                                       title = 'Sales price/sqaure foot and Gross Rent in San Francisco between 2010 and 2016',
                                       ylabel = 'Sales price per square foot/Gross Rent',
                                       xlabel = 'Year',
                                       grid = True
                                       
                                      )

:DynamicMap   [neighbourhood]
   :NdOverlay   [Variable]
      :Curve   [year]   (value)

### Step 6: Use the interactive visualisation to answer the following question:

**Question:** For the Anza Vista neighbourhood, is the average sale price per square foot for 2016 more or less than the price that’s listed for 2012? 

**Answer:** The sales price in 2016 is still higher than in 2012

---

## Build an Interactive Neighbourhood Map

For this part of the assignment, explore the geospatial relationships in the data by using interactive visualisations with hvPlot and GeoViews. To build your map, use the `sfo_data_df` DataFrame (created during the initial import), which includes the neighbourhood location data with the average prices. To do all this, complete the following steps:

1. Read the `neighbourhood_coordinates.csv` file from the `Resources` folder into the notebook, and create a DataFrame named `neighbourhood_locations_df`. Be sure to set the `index_col` of the DataFrame as “Neighbourhood”.

2. Using the original `sfo_data_df` Dataframe, create a DataFrame named `all_neighbourhood_info_df` that groups the data by neighbourhood. Aggregate the results by the `mean` of the group.

3. Review the two code cells that concatenate the `neighbourhood_locations_df` DataFrame with the `all_neighbourhood_info_df` DataFrame. Note that the first cell uses the [Pandas concat function](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.concat.html) to create a DataFrame named `all_neighbourhoods_df`. The second cell cleans the data and sets the “Neighbourhood” column. Be sure to run these cells to create the `all_neighbourhoods_df` DataFrame, which you’ll need to create the geospatial visualisation.

4. Using hvPlot with GeoViews enabled, create a `points` plot for the `all_neighbourhoods_df` DataFrame. Be sure to do the following:

    * Set the `size` parameter to “sale_price_sqr_foot”.

    * Set the `color` parameter to “gross_rent”.

    * Set the `frame_width` parameter to 700.

    * Set the `frame_height` parameter to 500.

    * Include a descriptive title.

    Note that your resulting plot should appear similar to the following image:

    ![A screenshot depicts an example of a scatter plot created with hvPlot and GeoViews.](Images/6-4-geoviews-plot.png)

5. Use the interactive map to answer the following question:

    * Which neighbourhood has the highest gross rent, and which has the highest sale price per square foot?

### Step 1: Read the `neighbourhood_coordinates.csv` file from the `Resources` folder into the notebook, and create a DataFrame named `neighbourhood_locations_df`. Be sure to set the `index_col` of the DataFrame as “Neighbourhood”.

In [138]:
# Load neighbourhoods coordinates data
neighbourhood_locations_df = pd.read_csv('./Resources/neighbourhoods_coordinates.csv', index_col='Neighbourhood')

# Review the DataFrame
neighbourhood_locations_df

Lat         Lon
Neighbourhood                            
Alamo Square        37.791012 -122.402100
Anza Vista          37.779598 -122.443451
Bayview             37.734670 -122.401060
Bayview Heights     37.728740 -122.410980
Bernal Heights      37.728630 -122.443050
...                       ...         ...
West Portal         37.740260 -122.463880
Western Addition    37.792980 -122.435790
Westwood Highlands  37.734700 -122.456854
Westwood Park       37.734150 -122.457000
Yerba Buena         37.792980 -122.396360

[73 rows x 2 columns]

### Step 2: Using the original `sfo_data_df` Dataframe, create a DataFrame named `all_neighbourhood_info_df` that groups the data by neighbourhood. Aggregate the results by the `mean` of the group.

In [140]:
# Calculate the mean values for each neighbourhood
all_neighbourhood_info_df = sfo_data_df.set_index('neighbourhood').drop('year', axis =1).groupby('neighbourhood').mean()

# Review the resulting DataFrame
# YOUR CODE HERE
all_neighbourhood_info_df.head(20)

sale_price_sqr_foot  housing_units   gross_rent
neighbourhood                                                                 
Alamo Square                            366.020712       378401.0  2817.285714
Anza Vista                              373.382198       379050.0  3031.833333
Bayview                                 204.588623       376454.0  2318.400000
Bayview Heights                         590.792839       382295.0  3739.000000
Bernal Heights                          576.746488       379374.5  3080.333333
Buena Vista Park                        452.680591       378076.5  2698.833333
Central Richmond                        394.422399       378401.0  2817.285714
Central Sunset                          423.687928       378401.0  2817.285714
Clarendon Heights                       487.244886       376454.0  2250.500000
Corona Heights                          587.539067       377232.8  2472.000000
Cow Hollow                              665.964042       378401.0  2817.285714
Croker Amazon                           303.004184       378076.5  2698.833333
Diamond Heights                         434.991739       375480.5  2016.000000
Downtown                                391.434378       378401.0  2817.285714
Duboce Triangle                         502.599156       378401.0  2780.250000
Eureka Valley/Dolores Heights           642.248671       378401.0  2817.285714
Excelsior                               388.765927       379050.0  3031.833333
Financial District North                391.362533       378401.0  2817.285714
Financial District South                455.836212       377232.8  2472.000000
Forest Knolls                           321.840837       374507.0  1781.500000

### Step 3: Review the two code cells that concatenate the `neighbourhood_locations_df` DataFrame with the `all_neighbourhood_info_df` DataFrame. 

Note that the first cell uses the [Pandas concat function](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.concat.html) to create a DataFrame named `all_neighbourhoods_df`. 

The second cell cleans the data and sets the “Neighbourhood” column. 

Be sure to run these cells to create the `all_neighbourhoods_df` DataFrame, which you’ll need to create the geospatial visualisation.

In [158]:
# Using the Pandas `concat` function, join the 
# neighbourhood_locations_df and the all_neighbourhood_info_df DataFrame
# The axis of the concatenation is "columns".
# The concat function will automatially combine columns with
# identical information, while keeping the additional columns.
all_neighbourhoods_df = pd.concat(
    [neighbourhood_locations_df, all_neighbourhood_info_df], 
    axis="columns",
    sort=True
).dropna()

# Review the resulting DataFrame
display(all_neighbourhoods_df.head(20))
display(all_neighbourhoods_df.tail(20))


Lat         Lon  sale_price_sqr_foot  \
Alamo Square                   37.791012 -122.402100           366.020712   
Anza Vista                     37.779598 -122.443451           373.382198   
Bayview                        37.734670 -122.401060           204.588623   
Bayview Heights                37.728740 -122.410980           590.792839   
Buena Vista Park               37.768160 -122.439330           452.680591   
Central Richmond               37.777890 -122.445170           394.422399   
Central Sunset                 37.749610 -122.489990           423.687928   
Clarendon Heights              37.753310 -122.447030           487.244886   
Corona Heights                 37.785530 -122.456000           587.539067   
Cow Hollow                     37.792980 -122.435790           665.964042   
Croker Amazon                  37.722800 -122.438690           303.004184   
Diamond Heights                37.728630 -122.443050           434.991739   
Duboce Triangle                37.769640 -122.426110           502.599156   
Eureka Valley/Dolores Heights  37.755540 -122.437790           642.248671   
Excelsior                      37.728740 -122.410980           388.765927   
Financial District North       37.791010 -122.402100           391.362533   
Financial District South       37.791010 -122.402100           455.836212   
Forest Knolls                  37.757060 -122.455090           321.840837   
Glen Park                      37.728630 -122.443050           623.826178   
Golden Gate Heights            37.757912 -122.464020           639.393557   

                               housing_units   gross_rent  
Alamo Square                        378401.0  2817.285714  
Anza Vista                          379050.0  3031.833333  
Bayview                             376454.0  2318.400000  
Bayview Heights                     382295.0  3739.000000  
Buena Vista Park                    378076.5  2698.833333  
Central Richmond                    378401.0  2817.285714  
Central Sunset                      378401.0  2817.285714  
Clarendon Heights                   376454.0  2250.500000  
Corona Heights                      377232.8  2472.000000  
Cow Hollow                          378401.0  2817.285714  
Croker Amazon                       378076.5  2698.833333  
Diamond Heights                     375480.5  2016.000000  
Duboce Triangle                     378401.0  2780.250000  
Eureka Valley/Dolores Heights       378401.0  2817.285714  
Excelsior                           379050.0  3031.833333  
Financial District North            378401.0  2817.285714  
Financial District South            377232.8  2472.000000  
Forest Knolls                       374507.0  1781.500000  
Glen Park                           378725.5  2899.500000  
Golden Gate Heights                 377622.2  2601.400000

Lat         Lon  sale_price_sqr_foot  \
Parkside                   37.749610 -122.489990           336.172661   
Parnassus/Ashbury Heights  37.763550 -122.457650           632.740454   
Portola                    37.734670 -122.401060           327.113925   
Potrero Hill               37.731960 -122.383240           662.013613   
Presidio Heights           37.785530 -122.456000           675.350212   
Russian Hill               37.792980 -122.435790           608.983217   
Silver Terrace             37.734670 -122.401060           170.292549   
South Beach                37.783230 -122.400650           650.124479   
South of Market            37.791010 -122.402100           570.271427   
Sunnyside                  37.734150 -122.457000           528.318332   
Telegraph Hill             37.800550 -122.404330           676.506578   
Twin Peaks                 37.753311 -122.447029           469.398626   
Union Square District      37.791010 -122.402100           903.993258   
Van Ness/ Civic Center     37.779510 -122.420220           404.150180   
Visitacion Valley          37.728740 -122.410980           301.466180   
West Portal                37.740260 -122.463880           498.488485   
Western Addition           37.792980 -122.435790           307.562201   
Westwood Highlands         37.734700 -122.456854           533.703935   
Westwood Park              37.734150 -122.457000           687.087575   
Yerba Buena                37.792980 -122.396360           576.709848   

                           housing_units   gross_rent  
Parkside                       377427.50  2555.166667  
Parnassus/Ashbury Heights      378401.00  2817.285714  
Portola                        376454.00  2318.400000  
Potrero Hill                   378401.00  2817.285714  
Presidio Heights               378401.00  2817.285714  
Russian Hill                   378401.00  2817.285714  
Silver Terrace                 380348.00  3528.000000  
South Beach                    375805.00  2099.000000  
South of Market                378401.00  2817.285714  
Sunnyside                      378725.50  2899.500000  
Telegraph Hill                 378401.00  2817.285714  
Twin Peaks                     378401.00  2817.285714  
Union Square District          377427.50  2555.166667  
Van Ness/ Civic Center         378401.00  2817.285714  
Visitacion Valley              381321.50  3657.000000  
West Portal                    376940.75  2515.500000  
Western Addition               377427.50  2555.166667  
Westwood Highlands             376454.00  2250.500000  
Westwood Park                  382295.00  3959.000000  
Yerba Buena                    377427.50  2555.166667

In [167]:
# Call the dropna function to remove any neighbourhoods that do not have data
all_neighbourhoods_df = all_neighbourhoods_df.reset_index().dropna()

# Rename the "index" column as "Neighbourhood" for use in the Visualisation
all_neighbourhoods_df = all_neighbourhoods_df.rename(columns={"index": "Neighbourhood"})

# Review the resulting DataFrame
display(all_neighbourhoods_df.head())
display(all_neighbourhoods_df.tail())

Neighbourhood        Lat         Lon  sale_price_sqr_foot  \
0      Alamo Square  37.791012 -122.402100           366.020712   
1        Anza Vista  37.779598 -122.443451           373.382198   
2           Bayview  37.734670 -122.401060           204.588623   
3   Bayview Heights  37.728740 -122.410980           590.792839   
4  Buena Vista Park  37.768160 -122.439330           452.680591   

   housing_units   gross_rent  
0       378401.0  2817.285714  
1       379050.0  3031.833333  
2       376454.0  2318.400000  
3       382295.0  3739.000000  
4       378076.5  2698.833333

Neighbourhood       Lat         Lon  sale_price_sqr_foot  \
64         West Portal  37.74026 -122.463880           498.488485   
65    Western Addition  37.79298 -122.435790           307.562201   
66  Westwood Highlands  37.73470 -122.456854           533.703935   
67       Westwood Park  37.73415 -122.457000           687.087575   
68         Yerba Buena  37.79298 -122.396360           576.709848   

    housing_units   gross_rent  
64      376940.75  2515.500000  
65      377427.50  2555.166667  
66      376454.00  2250.500000  
67      382295.00  3959.000000  
68      377427.50  2555.166667

In [183]:
print (all_neighbourhoods_df.sort_values('gross_rent'))
print (all_neighbourhoods_df.sort_values('sale_price_sqr_foot'))

         Neighbourhood       Lat         Lon  sale_price_sqr_foot  \
17       Forest Knolls  37.75706 -122.455090           321.840837   
11     Diamond Heights  37.72863 -122.443050           434.991739   
56         South Beach  37.78323 -122.400650           650.124479   
35       Miraloma Park  37.73415 -122.457000           779.810842   
66  Westwood Highlands  37.73470 -122.456854           533.703935   
..                 ...       ...         ...                  ...   
33      Merced Heights  37.71993 -122.465950           788.844818   
55      Silver Terrace  37.73467 -122.401060           170.292549   
63   Visitacion Valley  37.72874 -122.410980           301.466180   
3      Bayview Heights  37.72874 -122.410980           590.792839   
67       Westwood Park  37.73415 -122.457000           687.087575   

    housing_units  gross_rent  
17      374507.00     1781.50  
11      375480.50     2016.00  
56      375805.00     2099.00  
35      375967.25     2155.25  
66      376

### Step 4: Using hvPlot with GeoViews enabled, create a `points` plot for the `all_neighbourhoods_df` DataFrame. Be sure to do the following:

* Set the `geo` parameter to True.
* Set the `size` parameter to “sale_price_sqr_foot”.
* Set the `color` parameter to “gross_rent”.
* Set the `frame_width` parameter to 700.
* Set the `frame_height` parameter to 500.
* Include a descriptive title.

In [178]:
# Create a plot to analyse neighbourhood info
all_neighbourhoods_df.hvplot.points('Lon',
                                    'Lat',
                                    geo = True,
                                    tiles='OSM',
                                    size = 'sale_price_sqr_foot',
                                    color = 'gross_rent',
                                    frame_width = 700,
                                    frame_height = 500,
                                    colorbar= True,
                                    clabel = 'Gross Rent',
                                    alpha = 0.8,
                                    #groupby = 'Neighbourhood', #does not work as a function for some reason
                                    clim = (1000, 5000),
                                    title = "San Francisco price/rent plot",
                                    xlabel = 'Latitude',
                                    ylabel = 'Longitute'
                                    )


:Overlay
   .Tiles.I  :Tiles   [x,y]
   .Points.I :Points   [Lon,Lat]   (gross_rent,sale_price_sqr_foot)

### Step 5: Use the interactive map to answer the following question:

**Question:** Which neighbourhood has the highest gross rent, and which has the highest sale price per square foot?

**Answer:** Based on the hvplot provided above, the highest gross rent is located in Westwood park. To be entirely honest though, this map is not a good indicator of the highest rent. It is much easier to sort the values based on the gross rent and find the highest number.

The highest Sale Price is in Union Square District. Laso this is based on a calculation, rather than a graph.

## Compose Your Data Story

Based on the visualisations that you created, answer the following questions:

**Question:**  How does the trend in rental income growth compare to the trend in sales prices? Does this same trend hold true for all the neighbourhoods across San Francisco?

**Answer:** Generally speaking, it is true, the higher the sales price, the higher the rent price. However, by looking at the graph and the table, there are a few outliers that I would choose investing into, since they are cheaper, but bring the highest rent cash flow. For example, Silver Terrace is really cheap - only ~$170 per square foot, but it is in the top 5 of rental cash flow. 

On the other hand, Union Square District is the most expensive in terms of price, but it is not even in the top 10 of the rental cash flow income.

**Question:** What insights can you share with your company about the potential one-click, buy-and-rent strategy that they're pursuing? Do neighbourhoods exist that you would suggest for investment, and why?

**Answer:** I personally think that the analyses done on the data provided is insufficient to provide a one-click buy/rent option. If I were to revolutionize this idea, I would compare the price/rental income ratio, set the limit of the sales price for the customer, and let them buy the best deal - cheapest price, but highest rent. I might work on that in the future, but running out of time now.


Right now I would definitely invest in Silver Terrace, since it is one of the highest price to rent ratio.